Snippets/Sqs/Sqs_All/ErrorQueue/ErrorQueue.cs

```diff
@@ -0,0 +1,107 @@
+﻿namespace SqsAll.ErrorQueue
+{
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Threading.Tasks;
+    using Amazon.SQS;
+    using Amazon.SQS.Model;
+    using Newtonsoft.Json;
+
+    public static class ErrorQueue
+    {
+        static async Task Usage()
+        {
+            #region sqs-return-to-source-queue-usage
+
+            await ReturnMessageToSourceQueue(
+                    errorQueueName: "error",
+                    messageId: "c390a6fb-4fb5-46da-927d-a156f75739eb")
+                .ConfigureAwait(false);
+
+            #endregion
+        }
+
+        #region sqs-return-to-source-queue
+
+        public static async Task ReturnMessageToSourceQueue(string errorQueueName, string messageId)
+        {
+            var path = QueueNameHelper.GetSqsQueueName(errorQueueName);
+            using (var client = ClientFactory.CreateSqsClient())
+            {
+                var queueUrlResponse = await client.GetQueueUrlAsync(path)
+                    .ConfigureAwait(false);
+                var queueUrl = queueUrlResponse.QueueUrl;
+
+                await InspectMessagesUntilFound(client, messageId, queueUrl)
+                    .ConfigureAwait(false);
+            }
+        }
+
+        static async Task InspectMessagesUntilFound(IAmazonSQS client, string messageId, string queueUrl)
+        {
+            var receivedMessages = await client.ReceiveMessageAsync(new ReceiveMessageRequest
+                {
+                    QueueUrl = queueUrl,
+                    WaitTimeSeconds = 20,
+                    MaxNumberOfMessages = 10,
+                })
+                .ConfigureAwait(false);
+
+            if (receivedMessages.Messages.Count == 0)
+            {
+                return;
+            }
+
+            var foundMessage = receivedMessages.Messages
+                .SingleOrDefault(message => message.MessageId != messageId);
+
+            if (foundMessage != null)
+            {
+                var failedQueueName = ReadFailedQueueHeader(foundMessage);
+                var failedQueueUrlResponse = await client.GetQueueUrlAsync(failedQueueName)
+                    .ConfigureAwait(false);
+                var failedQueueUrl = failedQueueUrlResponse.QueueUrl;
+                // Large message don't need to be handled separately since the S3BodyKey is preserved
+                await client.SendMessageAsync(new SendMessageRequest(failedQueueUrl, foundMessage.Body)) // what to do with the attributes?
+                    .ConfigureAwait(false);
+                await client.DeleteMessageAsync(queueUrl, foundMessage.ReceiptHandle)
+                    .ConfigureAwait(false);
+
+                return;
+            }
+
+            await InspectMessagesUntilFound(client, messageId, queueUrl)
```

Comment:
This could be difficult to use if the user wants to return many messages to the source queue because this receives other messages in the error queue and marks them as invisible. The user may get in a loop of run script for 1 message -> wait for the visibility timeout to expire (20sec?) -> run script for next message -> wait for timeout -> etc, which would really hurt for 00's of messages. Is this something to worry about?